heartbeat/monitors/wrappers/monitorstate/monitorstate.go

```diff
@@ -0,0 +1,159 @@
+	Up         int         `json:"up"`
+	Down       int         `json:"down"`
+	// FlapHistory retains enough info so we can resume our flap
+	// computation if loading from ES or another source
+	FlapHistory []StateStatus `json:"flap_history"`
+	// Ends is a pointer to the prior state if this is the start of a new state
+	Ends *State `json:"ends"`
+	ctr  int
+}
+
+func (s *State) incrementCounters(status StateStatus) {
+	s.DurationMs = time.Since(s.StartedAt).Milliseconds()
+	s.Checks++
+	if status == StatusUp {
+		s.Up++
+	} else {
+		s.Down++
+	}
+}
+
+// truncate flap history to be at most as many items as the threshold indicates, minus one
+func (s *State) truncateFlapHistory() {
+	endIdx := len(s.FlapHistory)
+	if endIdx < 0 {
+		return // flap history is empty
+	}
+	// truncate to one less than the threshold since our later calculations
+	// an item that would stabilize the history at the threshold would start a new state
+	startIdx := endIdx - (FlappingThreshold - 1)
+	if startIdx <= 0 {
+		return
+	}
+	s.FlapHistory = s.FlapHistory[startIdx:endIdx]
+}
+
+// recordCheck updates the current state pointer to what the new state should be.
+// If the current state is continued it just updates counters and other record keeping,
+// if the state ends it actually swaps out the full value the state points to
+// and sets state.Ends.
+func (s *State) recordCheck(monitorID string, newStatus StateStatus) {
+	if s.Status == StatusFlapping {
+		s.truncateFlapHistory()
+
+		// Check if all statuses in flap history are identical, including the new status
+		hasStabilized := true
+		for _, histStatus := range s.FlapHistory {
+			if newStatus != histStatus {
+				hasStabilized = false
+				break
+			}
+		}
+
+		if !hasStabilized { // continue flapping
+			// Use the new flap history as part of the state
+			s.FlapHistory = append(s.FlapHistory, newStatus)
+			s.incrementCounters(newStatus)
+		} else { // flap has ended
+			s.transitionTo(monitorID, newStatus)
+		}
+	} else if s.Status == newStatus { // stable state, status has not changed
+		// The state is stable, no changes needed
+		s.incrementCounters(newStatus)
+	} else if s.Checks < FlappingThreshold {
+		// The state changed too quickly, we're now flapping
+		s.incrementCounters(newStatus)
+		s.Status = StatusFlapping
+		s.FlapHistory = append(s.FlapHistory, newStatus)
+	} else {
+		s.transitionTo(monitorID, newStatus)
+	}
+
+	// Ensure that the ends field is set to nil
+	// It's only needed on transitions
+	if s.Checks > 1 {
+		s.Ends = nil
+	}
+}
+
+func (s *State) transitionTo(monitorID string, newStatus StateStatus) {
+	// state has changed, but we aren't flapping (yet), since we've been stable past the
+	// flapping threshold
+	oldState := *s
+	*s = *newMonitorState(monitorID, newStatus, s.ctr)
+	// We don't need to retain extra data when transitioning
+	oldState.FlapHistory = nil
+	// W edon't want an infinite linked list!
+	oldState.Ends = nil
+	s.Ends = &oldState
+}
```

Comment:
One we determine a monitor is not flapping, I'm not sure it will be enough just to assign ends to the last state. The state needs to be adjusted to include the actual duration of the event.

Consider my above example, where a monitor is down for 5 minutes, but took 15 minutes to ensure it was not flapping. If the `ends` data is not updated, the duration for the state will report as 15 minutes, rather than 5 minutes.

Reply:
It's a good point to discuss, see my point earlier, I propose we disable flapping for now and discuss flapping in a follow-up PR. I'm apprehensive about an implementation where history can be rewritten, even the little bit we do now I don't like.